File: core/src/main/java/org/stepik/core/utils/Product.java
package org.stepik.core.utils;

import org.jetbrains.annotations.Nullable;

import java.util.Arrays;
import java.util.HashMap;
import java.util.Map;

/**
 * @author meanmail
 */
public enum Product {
    UNKNOWN_PRODUCT("Unknown"), IDEA_CE("IntelliJ IDEA Community Edition"), IDEA_ULTIMATE("IntelliJ IDEA"),
    PYCHARM_CE("PyCharm Community Edition"), PYCHARM_PRO("PyCharm");

    private static Map<String, Product> map;
    private final String name;


    Product(String name) {
        this.name = name;
    }

    public static Product of(@Nullable String name) {
        initMap();
        return map.getOrDefault(name, UNKNOWN_PRODUCT);
    }

    private static void initMap() {
        if (map == null) {
            map = new HashMap<>();
            Arrays.stream(values()).forEach(product -> map.put(product.getName(), product));
        }
    }

    public String getName() {
        return name;
    }

    @Override
    public String toString() {
        return getName();
    }
}

File: stepik-core/src/main/kotlin/org/stepik/core/utils/Product.kt
package org.stepik.core.utils

enum class Product(val product_name: String) {
    UNKNOWN_PRODUCT("Unknown"),
    IDEA_CE("IntelliJ IDEA Community Edition"),
    IDEA_ULTIMATE("IntelliJ IDEA"),
    PYCHARM_CE("PyCharm Community Edition"),
    PYCHARM_PRO("PyCharm");
    
    override fun toString() = product_name
    
    companion object {
        
        private val map: Map<String, Product> by lazy {
            values().associateBy { it.product_name }
        }
        
        fun of(product_name: String?) = map.getOrDefault(product_name, UNKNOWN_PRODUCT)
    }
}
